src/flow/src/batching_mode/engine.rs

```diff
@@ -0,0 +1,1151 @@
+fn create_table_with(plan: &LogicalPlan, sink_table_name: &str) -> Result<String, Error> {
+    let schema = plan.schema().fields();
+    let (first_time_stamp, all_pk_cols) = build_primary_key_constraint(plan, schema)?;
+
+    let mut schema_in_sql: Vec<String> = Vec::new();
+    for field in schema {
+        let ty = ConcreteDataType::from_arrow_type(field.data_type());
+        if first_time_stamp == Some(field.name().clone()) {
+            schema_in_sql.push(format!("\"{}\" {} TIME INDEX", field.name(), ty));
+        } else {
+            schema_in_sql.push(format!("\"{}\" {} NULL", field.name(), ty));
+        }
+    }
+    if first_time_stamp.is_none() {
+        schema_in_sql.push("\"update_at\" TimestampMillisecond default now()".to_string());
+        schema_in_sql.push(format!(
+            "\"{}\" TimestampMillisecond TIME INDEX default 0",
+            AUTO_CREATED_PLACEHOLDER_TS_COL
+        ));
+    } else {
+        schema_in_sql.push("\"update_at\" TimestampMillisecond default now()".to_string());
+    }
+
+    if !all_pk_cols.is_empty() {
+        schema_in_sql.push(format!("PRIMARY KEY ({})", all_pk_cols.iter().join(",")))
+    };
+
+    let col_defs = schema_in_sql.join(", ");
+
+    let create_table_sql = format!(
+        "CREATE TABLE IF NOT EXISTS {} ({})",
+        sink_table_name, col_defs
+    );
+
+    Ok(create_table_sql)
+}
+
+/// Return first timestamp column which is in group by clause and other columns which are also in group by clause
+///
+/// # Returns
+///
+/// * `Option<String>` - first timestamp column which is in group by clause
+/// * `Vec<String>` - other columns which are also in group by clause
+fn build_primary_key_constraint(
+    plan: &LogicalPlan,
+    schema: &Fields,
+) -> Result<(Option<String>, Vec<String>), Error> {
+    let mut pk_names = FindGroupByFinalName::default();
+
+    plan.visit(&mut pk_names)
+        .with_context(|_| DatafusionSnafu {
+            context: format!("Can't find aggr expr in plan {plan:?}"),
+        })?;
+
+    // if no group by clause, return empty
+    let pk_final_names = pk_names.get_group_expr_names().unwrap_or_default();
+    if pk_final_names.is_empty() {
+        return Ok((None, Vec::new()));
+    }
+
+    let all_pk_cols: Vec<_> = schema
+        .iter()
+        .filter(|f| pk_final_names.contains(f.name()))
+        .map(|f| f.name().clone())
+        .collect();
+    // auto create table use first timestamp column in group by clause as time index
+    let first_time_stamp = schema
+        .iter()
+        .find(|f| {
+            all_pk_cols.contains(&f.name().clone())
+                && ConcreteDataType::from_arrow_type(f.data_type()).is_timestamp()
+        })
+        .map(|f| f.name().clone());
+
+    let all_pk_cols: Vec<_> = all_pk_cols
+        .into_iter()
+        .filter(|col| first_time_stamp != Some(col.to_string()))
+        .collect();
+
+    Ok((first_time_stamp, all_pk_cols))
+}
+
+#[derive(Debug)]
+pub struct TaskState {
```

Comment:
This file/mod is too large and should be split into `task`, `engine`, and `state`.